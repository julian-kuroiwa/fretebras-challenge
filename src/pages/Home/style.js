import styled, { css } from 'styled-components';
import { shade } from 'polished';

import { mediaQueries } from '../../styles/media';

import LoaderComponent from '../../components/Loader';

import BurguerBg from '../../assets/burguer-background.jpg';
import BurguerImg from '../../assets/burguer.jpg';

export const Hero = styled.section`
  display: flex;
  flex-direction: column;
  justify-content: center;
  min-height: 600px;
  background: url(${BurguerBg}) no-repeat left center;
  background-size: cover;
  box-shadow: inset 0px 120px 140px #000000;

  .content {
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    max-width: 550px;

    ${mediaQueries('small')`
        display: block;
        flex-direction: initial;
        justify-content: initial;
        align-items: initial;
      `};

    h1 {
      font-size: 32px;
      color: #fff;
      line-height: 36px;
      margin-bottom: 40px;
      text-align: center;

      ${mediaQueries('small')`
        font-size: 60px;
        line-height: 60px;
        text-align: left;
      `};
    }

    p {
      font-size: 18px;
      color: #fff;
      line-height: 25px;
      margin-bottom: 40px;
      text-align: center;

      ${mediaQueries('small')`
        font-size: 24px;
        line-height: 36px;
        text-align: left;
      `};
    }
  }
`;

export const Info = styled.section`
  display: flex;
  align-items: center;
  min-height: 600px;
  background-color: #f6efdf;
  padding: 50px 0;

  .content {
    display: flex;
    flex-direction: column;
    align-items: center;

    ${mediaQueries('small')`
      flex-direction: row;
      justify-content: space-between;
    `};

    > div {
      display: flex;
      flex-direction: column;
      align-items: center;
      justify-content: center;
      margin-bottom: 50px;

      ${mediaQueries('small')`
        max-width: 50%;
        display: block;
        flex-direction: row;
        align-items: initial;
        justify-content: initial;
        margin-bottom: 0;
      `};

      h2 {
        font-size: 28px;
        color: #f74a3e;
        line-height: 36px;
        margin-bottom: 40px;
        text-align: center;

        ${mediaQueries('small')`
          font-size: 60px;
          line-height: 60px;
          text-align: left;
        `};
      }

      p {
        font-size: 18px;
        color: #f74a3e;
        line-height: 25px;
        margin-bottom: 40px;
        text-align: center;

        ${mediaQueries('small')`
        font-size: 24px;
        line-height: 36px;
        text-align: left;
      `};
      }
    }

    .round-image {
      background: url(${BurguerImg}) no-repeat 59% 20%;
      background-size: 210%;
      width: 300px;
      height: 300px;
      border-radius: 50%;

      ${mediaQueries('small')`
        width: 400px;
        height: 400px;
      `};
    }
  }
`;

export const Search = styled.section`
  display: flex;
  align-items: center;
  min-height: 600px;
  background-color: #f74a3e;

  .content {
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    max-width: 650px;
    width: 100%;
    margin: 0 auto;

    h2 {
      font-size: 28px;
      color: #fff;
      line-height: 36px;
      margin-bottom: 40px;
      text-align: center;

      ${mediaQueries('small')`
          font-size: 40px;
          line-height: 40px;
        `};
    }

    p {
      font-size: 18px;
      color: #fff;
      line-height: 25px;
      margin-bottom: 100px;
      text-align: center;

      ${mediaQueries('small')`
        font-size: 24px;
        line-height: 36px;
      `};
    }

    form {
      width: 100%;
      display: flex;
      align-items: center;
      flex-direction: column;
    }
  }
`;

export const Popular = styled.section`
  display: flex;
  align-items: center;
  min-height: 600px;
  background-color: #f6efdf;
  padding: 50px 0;

  .content {
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    width: 100%;
    margin: 0 auto;

    h2 {
      font-size: 28px;
      color: #f74a3e;
      line-height: 36px;
      margin-bottom: 80px;
      text-align: center;

      ${mediaQueries('small')`
          font-size: 40px;
          line-height: 40px;
      `};
    }

    > div {
      : none;
      flex-direction: column;
      display: flex;
      justify-content: space-between;

      ${mediaQueries('medium')`
        flex-direction: row;
      `};
    }
  }
`;

export const Autocomplete = styled.div`
  width: 100%;
  margin-bottom: 40px;
  position: relative;

  .input-container {
    width: 100%;
    position: relative;

    input {
      width: 100%;
      border: 2px solid #fff;
      border-radius: 50px;
      height: 70px;
      padding: 0 20px 0 70px;

      ${(props) =>
        props.error &&
        css`
          border-color: ${shade(0.2, '#ff0090')};
        `}
    }

    i {
      position: absolute;
      color: #f74a3e;
      top: 50%;
      left: 20px;
      font-size: 35px;
      transform: translateY(-50%);
    }
  }

  .results-container {
    background-color: #fff;
    : none;
    max-width: 80%;
    position: absolute;
    top: 100%;
    left: 50%;
    transform: translate(-50%, 5px);
    width: 100%;
    max-height: 200px;
    overflow: scroll;

    > button {
      padding: 20px;
      border: 0;
      display: block;
      width: 100%;
      text-align: left;
      text-transform: uppercase;

      &:hover {
        background-color: ${shade(0.2, '#fff')};
      }
    }
  }

  > span {
    display: block;
    width: 100%;
    margin-top: 5px;
    text-align: center;
    color: #fff;
    font-weight: bold;
    position: absolute;
    top: 100%;
    left: 0;
  }
`;

export const Loader = styled(LoaderComponent)`
  position: absolute;
  top: 50%;
  right: 20px;
  transform: translateY(-50%);
  background: #fff;
`;
