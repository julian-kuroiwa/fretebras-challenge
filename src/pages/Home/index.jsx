import React, { useState, useEffect, useRef } from 'react';
import { useHistory } from 'react-router-dom';

import Header from '../../components/Header';
import Card from '../../components/Card';
import Wrapper from '../../components/Wrapper';
import Button from '../../components/Button';
import Footer from '../../components/Footer';

import { Hero, Info, Search, Popular, Autocomplete, Loader } from './style';
import api from '../../services/api';

import useDebounce from '../../hooks/useDebounce';

const Home = () => {
  const history = useHistory();
  const inputRef = useRef(null);

  const [inputValue, setInputValue] = useState('');
  const [isLoading, setIsLoading] = useState(false);
  const [results, setResults] = useState([]);
  const [citySelected, setCitySelected] = useState({});
  const [error, setError] = useState('');

  const popularPackage = [
    {
      id: '1',
      name: 'Package I',
      price: '$10.00',
      description:
        'Lorem ipsum dolor sit amet consectetur adipisicing elit. Fugit, perferendis?',
    },
    {
      id: '2',
      name: 'Package II',
      price: '$20.00',
      description:
        'Lorem ipsum dolor sit amet consectetur adipisicing elit. Fugit, perferendis?',
    },
    {
      id: '3',
      name: 'Package III',
      price: '$30.00',
      description:
        'Lorem ipsum dolor sit amet consectetur adipisicing elit. Fugit, perferendis?',
    },
  ];

  const debounceSearch = useDebounce(inputValue, 1000);

  useEffect(() => {
    if (debounceSearch) {
      setIsLoading(true);
      setError('');

      api
        .get(`/cities?q=${debounceSearch}`)
        .then((response) => {
          setIsLoading(false);

          if (!response.data.location_suggestions.length) {
            setError('No results found');

            return;
          }

          setResults(response.data.location_suggestions);
        })
        .catch((err) => {
          setIsLoading(false);
          setError('Error trying to find the city, please try again!');
        });

      return;
    }

    setResults([]);
  }, [debounceSearch]);

  const changeHandle = (event) => {
    setInputValue(event.target.value);
    setError('');
  };

  const clickOptionHandle = (id, name) => {
    setCitySelected({ id, name });
    setResults([]);
  };

  const submitHandle = (event) => {
    event.preventDefault();

    if (!citySelected.name && !citySelected.id) {
      setError('Search a city before proceed');

      return;
    }

    history.push(`/restaurants/${citySelected.name}/${citySelected.id}`);
  };

  return (
    <>
      <Header />
      <Hero>
        <Wrapper>
          <div className="content">
            <h1>Get cashback up to 50%</h1>
            <p>
              Lorem ipsum dolor sit amet consectetur, adipisicing elit. Deleniti
              incidunt, quae suscipit eligendi voluptatum unde.
            </p>
            <Button type="button">Order Now</Button>
          </div>
        </Wrapper>
      </Hero>
      <Info>
        <Wrapper>
          <div className="content">
            <div>
              <h2>Big Burguer</h2>
              <p>
                Lorem ipsum dolor sit amet consectetur, adipisicing elit.
                Deleniti incidunt, quae suscipit eligendi voluptatum unde.
              </p>
              <Button type="button">Order Now</Button>
            </div>
            <div className="round-image" />
          </div>
        </Wrapper>
      </Info>
      <Search>
        <Wrapper>
          <div className="content">
            <h2>Search the best burguers</h2>
            <p>
              Lorem ipsum dolor sit amet consectetur, adipisicing elit. Deleniti
              incidunt, quae suscipit eligendi voluptatum unde.
            </p>
            <form onSubmit={submitHandle}>
              <Autocomplete error={!!error}>
                <div className="input-container">
                  <input
                    type="search"
                    placeholder="ex. SÃ£o Paulo"
                    ref={inputRef}
                    onChange={changeHandle}
                  />
                  <i className="fas fa-map-marker-alt" />
                  {isLoading && <Loader />}
                </div>
                {error && <span>{error}</span>}
                {results.length > 0 && (
                  <div className="results-container">
                    {results.map((result) => (
                      <button
                        type="button"
                        key={result.id}
                        onClick={() =>
                          clickOptionHandle(result.id, result.name)
                        }
                      >
                        {result.name}
                      </button>
                    ))}
                  </div>
                )}
              </Autocomplete>
              <Button type="submit" isNegative>
                Search
              </Button>
            </form>
          </div>
        </Wrapper>
      </Search>
      <Popular>
        <Wrapper>
          <div className="content">
            <h2>Popular Package</h2>
            <div>
              {popularPackage.map((item) => (
                <Card key={item.id} item={item} />
              ))}
            </div>
          </div>
        </Wrapper>
      </Popular>
      <Footer />
    </>
  );
};

export default Home;
