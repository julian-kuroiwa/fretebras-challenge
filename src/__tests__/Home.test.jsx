import { shallow } from 'enzyme';
import React from 'react';

import Home from '../pages/Home';

jest.mock('react-router', () => ({
  useHistory: () => ({ push: jest.fn() }),
}));

describe('Home', () => {
  it('should match snapshot', () => {
    const component = shallow(<Home />);

    expect(component).toMatchSnapshot();
  });

  it('should contain an form', () => {
    const component = shallow(<Home />);

    const form = component.find('form');

    expect(form).toBeTruthy();
    expect(form.find('input')).toBeTruthy();
    expect(form.find('button')).toBeTruthy();
  });
});
