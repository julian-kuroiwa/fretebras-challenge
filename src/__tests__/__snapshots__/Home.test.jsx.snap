// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home should match snapshot 1`] = `
<Fragment>
  <Header />
  <styled.section>
    <Wrapper>
      <div
        className="content"
      >
        <h1>
          Get cashback up to 50%
        </h1>
        <p>
          Lorem ipsum dolor sit amet consectetur, adipisicing elit. Deleniti incidunt, quae suscipit eligendi voluptatum unde.
        </p>
        <Button
          isNegative={false}
          type="button"
        >
          Order Now
        </Button>
      </div>
    </Wrapper>
  </styled.section>
  <styled.section>
    <Wrapper>
      <div
        className="content"
      >
        <div>
          <h2>
            Big Burguer
          </h2>
          <p>
            Lorem ipsum dolor sit amet consectetur, adipisicing elit. Deleniti incidunt, quae suscipit eligendi voluptatum unde.
          </p>
          <Button
            isNegative={false}
            type="button"
          >
            Order Now
          </Button>
        </div>
        <div
          className="round-image"
        />
      </div>
    </Wrapper>
  </styled.section>
  <styled.section>
    <Wrapper>
      <div
        className="content"
      >
        <h2>
          Search the best burguers
        </h2>
        <p>
          Lorem ipsum dolor sit amet consectetur, adipisicing elit. Deleniti incidunt, quae suscipit eligendi voluptatum unde.
        </p>
        <form
          onSubmit={[Function]}
        >
          <styled.div
            error={false}
          >
            <div
              className="input-container"
            >
              <input
                onChange={[Function]}
                placeholder="ex. SÃ£o Paulo"
                type="search"
              />
              <i
                className="fas fa-map-marker-alt"
              />
            </div>
          </styled.div>
          <Button
            isNegative={true}
            type="submit"
          >
            Search
          </Button>
        </form>
      </div>
    </Wrapper>
  </styled.section>
  <styled.section>
    <Wrapper>
      <div
        className="content"
      >
        <h2>
          Popular Package
        </h2>
        <div>
          <Card
            item={
              Object {
                "description": "Lorem ipsum dolor sit amet consectetur adipisicing elit. Fugit, perferendis?",
                "id": "1",
                "name": "Package I",
                "price": "$10.00",
              }
            }
            key="1"
          />
          <Card
            item={
              Object {
                "description": "Lorem ipsum dolor sit amet consectetur adipisicing elit. Fugit, perferendis?",
                "id": "2",
                "name": "Package II",
                "price": "$20.00",
              }
            }
            key="2"
          />
          <Card
            item={
              Object {
                "description": "Lorem ipsum dolor sit amet consectetur adipisicing elit. Fugit, perferendis?",
                "id": "3",
                "name": "Package III",
                "price": "$30.00",
              }
            }
            key="3"
          />
        </div>
      </div>
    </Wrapper>
  </styled.section>
  <Footer />
</Fragment>
`;
